File: netlogo-gui/src/main/window/RuntimeErrorDialog.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.api.LogoException;
import org.nlogo.workspace.AbstractWorkspace;

import java.util.ArrayList;
import java.util.List;

public strictfp class RuntimeErrorDialog
    extends org.nlogo.swing.MessageDialog {

  private String title;
  private boolean ordinaryError;
  private boolean askForBugReport;
  private String errorMessage;
  private String javaStackTrace;
  private String eventTrace;
  private final String className;
  private String threadName;

  private String textWithoutDetails;
  private String textWithDetails;

  private boolean suppressJavaExceptionDialogs = false;
  private String modelName;

  private static final javax.swing.JButton SUPPRESS_BUTTON =
      new javax.swing.JButton("Don't show again");
  private static final javax.swing.JCheckBox CHECKBOX =
      new javax.swing.JCheckBox("Show internal details");

  public static org.nlogo.swing.MessageDialog init(java.awt.Component owner) {
    dialog = new RuntimeErrorDialog(owner);
    return dialog;
  }

  public static void deactivate() {
    dialog = null;
  }

  public static boolean alreadyVisible() {
    return dialog != null && dialog.isVisible();
  }

  public static void show(String title,
                          org.nlogo.nvm.Context context,
                          org.nlogo.nvm.Instruction instruction,
                          Thread thread,
                          Throwable throwable) {
    if (dialog != null) {
      ((RuntimeErrorDialog) dialog)
          .doShow(title, context, instruction, thread, throwable);
    }
  }

  public static void setModelName(String name) {
    ((RuntimeErrorDialog) dialog).modelName = name;
  }

  public static boolean suppressJavaExceptionDialogs() {
    return ((RuntimeErrorDialog) dialog).suppressJavaExceptionDialogs;
  }

  ///

  private void doShow(String title,
                      org.nlogo.nvm.Context context,
                      org.nlogo.nvm.Instruction instruction,
                      Thread thread,
                      Throwable throwable) {
    this.title = title;
    threadName = thread.getName();
    ordinaryError = throwable instanceof LogoException;
    // we don't need bug reports on known issues like OutOfMemoryError - ST 4/29/10
    askForBugReport = !(ordinaryError || hasKnownAncestorCause(throwable));
    // context is only non-null if the exception actually occured inside
    // running Logo code (and not, for example, in the GUI)
    SUPPRESS_BUTTON.setVisible(!ordinaryError && context == null);
    javaStackTrace = org.nlogo.util.Utils.getStackTrace(throwable);
    eventTrace = org.nlogo.window.Event.recentEventTrace();
    if (context != null) {
      errorMessage = context.buildRuntimeErrorMessage(instruction, throwable);
    } else if (ordinaryError) {
      errorMessage = throwable.getMessage();
    } else {
      errorMessage = null;
    }
    buildTexts();
    // don't show spurious dialogs caused by
    // http://bugs.sun.com/view_bug.do?bug_id=6828938
    // http://bugs.sun.com/view_bug.do?bug_id=6899297
    // http://bugs.sun.com/view_bug.do?bug_id=6857057
    // the underlying issue was fixed as of Java 6 Update 20, but we might
    // as well still be careful
    // - ST 9/12/09
    if (textWithDetails.indexOf("sun.font.FontDesignMetrics.charsWidth") == -1) {
      showJavaDetails(CHECKBOX.isSelected());
    }
  }

  private boolean hasKnownAncestorCause(Throwable t) {
    return t instanceof java.lang.OutOfMemoryError ||
        (t.getCause() != null && hasKnownAncestorCause(t.getCause()));
  }

  protected RuntimeErrorDialog(java.awt.Component owner) {
    super(owner);
    className = owner.getClass().getName();
  }

  @Override
  protected List<javax.swing.JComponent> makeButtons() {
    List<javax.swing.JComponent> buttons =
        new ArrayList<javax.swing.JComponent>();
    buttons.addAll(super.makeButtons());
    if (!AbstractWorkspace.isApplet()) {
      javax.swing.JButton copyButton = new javax.swing.JButton("Copy");
      copyButton.addActionListener
          (new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent e) {
              textArea.selectAll();
              textArea.copy();
              textArea.setCaretPosition(0);
            }
          });
      buttons.add(copyButton);
    }
    CHECKBOX.addItemListener
        (new java.awt.event.ItemListener() {
          public void itemStateChanged(java.awt.event.ItemEvent e) {
            showJavaDetails(CHECKBOX.isSelected());
          }
        });
    buttons.add(CHECKBOX);
    SUPPRESS_BUTTON.addActionListener
        (new java.awt.event.ActionListener() {
          public void actionPerformed(java.awt.event.ActionEvent e) {
            suppressJavaExceptionDialogs = true;
            setVisible(false);
          }
        });
    buttons.add(SUPPRESS_BUTTON);
    return buttons;
  }

  private void buildTexts() {
    String extra = "";
    if (askForBugReport) {
      extra = "NetLogo is unable to supply you with more details about this error.  Please report the problem\n"
          + "at https://github.com/NetLogo/NetLogo/issues, or to bugs@ccl.northwestern.edu, and paste the\n"
          + "contents of this window into your report.\n\n";
    }
    textWithoutDetails = "";
    if (errorMessage != null) {
      textWithoutDetails = errorMessage;
    }
    textWithDetails = textWithoutDetails;
    if (!textWithDetails.equals("")) {
      textWithDetails += "\n\n";
    }
    textWithDetails +=
        extra + javaStackTrace + "\n" +
            org.nlogo.api.Version.version() + "\n" +
            "main: " + className + "\n" +
            "thread: " + threadName + "\n" +
            org.nlogo.util.SysInfo.getVMInfoString() + "\n" +
            org.nlogo.util.SysInfo.getOSInfoString() + "\n" +
            org.nlogo.util.SysInfo.getScalaVersionString() + "\n" +
            org.nlogo.util.SysInfo.getJOGLInfoString() + "\n" +
            org.nlogo.util.SysInfo.getGLInfoString() + "\n" +
            "model: " + modelName + "\n\n" +
            eventTrace;
    if (!AbstractWorkspace.isApp()) {
      String browser = org.nlogo.util.SysInfo.getBrowserInfoString();
      textWithDetails += "\n";
      textWithDetails +=
          (browser == null)
              ? "running as applet in unknown browser"
              : ("browser info: " + browser);
    }
  }

  private void showJavaDetails(boolean flag) {
    if (ordinaryError) {
      CHECKBOX.setVisible(true);
    } else {
      CHECKBOX.setVisible(false);
      flag = true;
    }
    int lines = 1;
    int lineBegin = 0;
    int longestLine = 0;
    String text =
        flag
            ? textWithDetails
            : textWithoutDetails;
    for (int i = 0; i < text.length(); i++) {
      char c = text.charAt(i);
      if (c == '\n' || c == '\r') {
        lines++;
        if (i - lineBegin > longestLine) {
          longestLine = i - lineBegin;
        }
        lineBegin = i;
      }
    }
    longestLine += 2; // just in case
    if (lines < 5) {
      lines = 5;
    }
    if (lines > 15) {
      lines = 15;
    }
    if (longestLine > 70) {
      longestLine = 70;
    }
    ((RuntimeErrorDialog) dialog)
        .doShow(title, text, lines, longestLine);
  }

}

File: netlogo-gui/src/main/window/RuntimeErrorDialog.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window

import java.awt.Component
import java.awt.event.{ ActionEvent, ActionListener, ItemEvent, ItemListener }
import java.util.{ ArrayList => JArrayList }
import java.util.{ List => JList }
import java.lang.OutOfMemoryError

import javax.swing.{ JButton, JCheckBox, JComponent }

import org.nlogo.core.I18N
import org.nlogo.api.{ LogoException, Version }
import org.nlogo.nvm.{ Context, Instruction }
import org.nlogo.workspace.AbstractWorkspace
import org.nlogo.swing.{ BrowserLauncher, MessageDialog }
import org.nlogo.util.Utils
import org.nlogo.util.SysInfo

import scala.annotation.tailrec
import scala.collection.JavaConverters._

case class ErrorInfo(var throwable: Throwable, var context: Option[Context] = None, var instruction: Option[Instruction] = None) {
  def ordinaryError: Boolean = throwable.isInstanceOf[LogoException]

  def hasKnownCause: Boolean = knownAncestorCause(throwable)

  def isOutOfMemory: Boolean = knownAncestorCause(throwable)

  def hasContext: Boolean = context.nonEmpty

  def errorMessage: Option[String] =
    context.flatMap(c => instruction.map(i => (c, i))).map {
      case (ctx, ins) => ctx.buildRuntimeErrorMessage(ins, throwable)
    } orElse (if (ordinaryError) Some(throwable.getMessage) else None)

  @tailrec
  private def knownAncestorCause(t: Throwable): Boolean =
    t.isInstanceOf[OutOfMemoryError] || (t.getCause != null && knownAncestorCause(t.getCause))
}

case class DebuggingInfo(var className: String, var threadName: String, var modelName: String, var eventTrace: String, var javaStackTrace: String) {
  def debugInfo =
    s"""|${Version.version}
        |main: $className
        |thread: $threadName
        |${SysInfo.getVMInfoString}
        |${SysInfo.getOSInfoString}
        |${SysInfo.getScalaVersionString}
        |${SysInfo.getJOGLInfoString}
        |${SysInfo.getGLInfoString}
        |model: $modelName""".stripMargin

  def detailedInformation: String = s"""|$javaStackTrace
                                        |$debugInfo
                                        |
                                        |$eventTrace""".stripMargin
}

object RuntimeErrorDialog {
  val PleaseReportText = I18N.gui.get("error.dialog.pleaseReport")

  private var debuggingInfo = DebuggingInfo("", "", "", "", "")
  private var errorInfo = ErrorInfo(null)
  private var unknownDialog: Option[UnknownErrorDialog]  = Option.empty[UnknownErrorDialog]
  private var logoDialog:    Option[LogoExceptionDialog] = Option.empty[LogoExceptionDialog]
  private var memoryDialog:  Option[OutOfMemoryDialog]   = Option.empty[OutOfMemoryDialog]

  def init(owner: Component): Unit = {
    debuggingInfo.className = owner.getClass.getName

    unknownDialog = Some(new UnknownErrorDialog(owner))
    logoDialog    = Some(new LogoExceptionDialog(owner))
    memoryDialog  = Some(new OutOfMemoryDialog(owner))
  }

  def setModelName(name: String): Unit = {
    debuggingInfo.modelName = name
  }

  def deactivate(): Unit = {
    unknownDialog = None
    logoDialog    = None
    memoryDialog  = None
    debuggingInfo = DebuggingInfo("", "", "", "", "")
    errorInfo = ErrorInfo(null)
  }

  def alreadyVisible: Boolean = {
    Seq(unknownDialog, logoDialog, memoryDialog).flatten.exists(_.isVisible)
  }

  def show(context: Context, instruction: Instruction, thread: Thread, throwable: Throwable): Unit = {
      debuggingInfo.threadName     = thread.getName
      debuggingInfo.eventTrace     = Event.recentEventTrace()
      debuggingInfo.javaStackTrace = Utils.getStackTrace(throwable)
      errorInfo.throwable   = throwable
      errorInfo.context     = Option(context)
      errorInfo.instruction = Option(instruction)
      throwable match {
        case l: LogoException             => logoDialog.foreach(_.doShow(errorInfo, debuggingInfo))
        case _ if errorInfo.isOutOfMemory => memoryDialog.foreach(_.doShow())
        case _                            => unknownDialog.foreach(_.doShow("Internal Error", errorInfo, debuggingInfo))
      }
  }

  def suppressJavaExceptionDialogs: Boolean = {
    unknownDialog.map(_.suppressJavaExceptionDialogs).getOrElse(false)
  }
}

import RuntimeErrorDialog._

trait CopyButton {
  def copy(): Unit

  lazy val copyButton: JButton = new JButton(I18N.gui.get("menu.edit.copy"))
  copyButton.addActionListener(new ActionListener() {
    def actionPerformed(e: ActionEvent): Unit = { copy() }
  })
}

trait RuntimeErrorDialog {
  protected var textWithDetails: String    = ""
  protected var textWithoutDetails: String = ""

  protected lazy val checkbox = {
    val b = new JCheckBox(I18N.gui.get("error.dialog.showInternals"))
    b.addItemListener(new ItemListener() {
      def itemStateChanged(e: ItemEvent): Unit = {
        showJavaDetails(b.isSelected)
      }
    })
    b
  }

  protected def showJavaDetails(showDetails: Boolean): Unit = {
    var lines = 1
    var lineBegin = 0
    var longestLine = 0
    var text =
        if (showDetails) textWithDetails else textWithoutDetails

    var i = 0
    while (i < text.length) {
      text.charAt(i) match {
        case '\n' | '\r' =>
          lines += 1
          longestLine = longestLine max (i - lineBegin)
          lineBegin = i
        case _ =>
      }
      i += 1
    }

    longestLine += 2 // pad
    lines = lines.max(5).min(15)
    longestLine = longestLine.min(70)

    showText(text, lines, longestLine)
  }

  protected def showText(text: String, rows: Int, columns: Int): Unit
}

class UnknownErrorDialog(owner: Component) extends MessageDialog(owner) with RuntimeErrorDialog with CopyButton {
  private lazy val suppressButton   = new JButton(I18N.gui.get("error.dialog.suppress"))

  private var dialogTitle: String = ""
  private var errorHeader: String = ""

  var suppressJavaExceptionDialogs: Boolean = false

  def doShow(showTitle: String, errorInfo: ErrorInfo, debuggingInfo: DebuggingInfo): Unit = {
    dialogTitle = showTitle
    // we don't need bug reports on known issues like OutOfMemoryError - ST 4/29/10
    errorHeader =
      if (! (errorInfo.ordinaryError || errorInfo.hasKnownCause)) PleaseReportText
      else ""
    // only has context if the exception occured inside running Logo code (and not, for example, in the GUI)
    suppressButton.setVisible(! (errorInfo.ordinaryError || errorInfo.hasContext))
    buildTexts(errorInfo, debuggingInfo)
    checkbox.setVisible(errorInfo.ordinaryError)
    showJavaDetails(! errorInfo.ordinaryError || checkbox.isSelected)
  }

  override def copy(): Unit = {
    val beginIndex = textArea.getText.indexOf(errorHeader) + errorHeader.length
    textArea.select(beginIndex, textArea.getText.length)
    textArea.copy()
    textArea.setCaretPosition(0)
  }

  override protected def makeButtons(): JList[JComponent] = {
    val buttons: JList[JComponent] = new JArrayList[JComponent]()
    buttons.addAll(super.makeButtons())
    buttons.add(copyButton)
    buttons.add(checkbox)
    suppressButton.addActionListener(new ActionListener() {
      def actionPerformed(e: ActionEvent): Unit = {
        suppressJavaExceptionDialogs = true
        setVisible(false)
      }
    })
    buttons.add(suppressButton)
    buttons
  }

  private def buildTexts(errorInfo: ErrorInfo, debuggingInfo: DebuggingInfo): Unit = {
    val detailedInformation =
      s"$errorHeader\n${debuggingInfo.detailedInformation}"
    textWithoutDetails = errorInfo.errorMessage.getOrElse("")
    textWithDetails = errorInfo.errorMessage
      .map(_ + "\n\n" + detailedInformation)
      .getOrElse(detailedInformation)
  }

  override protected def showText(text: String, rows: Int, columns: Int): Unit =
    doShow(dialogTitle, text, rows, columns)
}

class LogoExceptionDialog(owner: Component) extends MessageDialog(owner) with RuntimeErrorDialog with CopyButton {
  private val dialogTitle: String = "Runtime Error"

  private var errorMessage: Option[String] = None

  def doShow(errorInfo: ErrorInfo, debuggingInfo: DebuggingInfo): Unit = {
    buildTexts(errorInfo, debuggingInfo)
    showJavaDetails(checkbox.isSelected)
  }

  override def copy(): Unit = {
    textArea.select(0, textArea.getText.length)
    textArea.copy()
    textArea.setCaretPosition(0)
  }

  override protected def makeButtons(): JList[JComponent] = {
    val buttons: JList[JComponent] = new JArrayList[JComponent]()
    buttons.addAll(super.makeButtons())
    buttons.add(copyButton)
    buttons.add(checkbox)
    buttons
  }

  private def buildTexts(errorInfo: ErrorInfo, debuggingInfo: DebuggingInfo): Unit = {
    textWithoutDetails = errorInfo.errorMessage.getOrElse("")
    textWithDetails    = errorInfo.errorMessage
      .map(_ + "\n\n" + debuggingInfo.detailedInformation)
      .getOrElse(debuggingInfo.detailedInformation)
  }

  override protected def showText(text: String, rows: Int, columns: Int): Unit =
    doShow(dialogTitle, text, rows, columns)
}

class OutOfMemoryDialog(owner: Component) extends MessageDialog(owner) with RuntimeErrorDialog {
  private val dialogTitle: String = "Model Too Large"
  private val ErrorText = I18N.gui.get("error.dialog.outOfMemory")

  override protected def makeButtons(): JList[JComponent] = {
    val buttons: JList[JComponent] = new JArrayList[JComponent]()
    buttons.addAll(super.makeButtons())
    val openFAQ = new JButton(I18N.gui.get("error.dialog.openFAQ"))
    val baseFaqUrl: String = {
      val docRoot = System.getProperty("netlogo.docs.dir", "docs")
      docRoot + "/faq.html"
    }
    openFAQ.addActionListener(new ActionListener() {
      def actionPerformed(e: ActionEvent): Unit = {
        BrowserLauncher.openURL(owner, baseFaqUrl, "#howbig", true)
      }
    })
    buttons.add(openFAQ)
    buttons
  }

  def doShow(): Unit = {
    textWithDetails = ErrorText
    showJavaDetails(true)
  }

  override protected def showText(text: String, rows: Int, columns: Int): Unit =
    doShow(dialogTitle, text, rows, columns)
}
